### PR TITLE
Generalize callbacks

### LINKED ISSUE #482: Generalize callbacks
Currently, we can only use monthly callbacks. This tasks generalizes the current funcitonality to arbitrary time intervals and functions. This is currently applied to model state checkpointing and updating the logger for the first day of the month for BCReader. 

# Components
- [x] implementation in src
- [x] unit tests
- [x] implementation in driver
- [x] documantation

# QA
- [ ] no behaviour change

### PR DESCRIPTION
**Pull request by LenkaNovak**
## Purpose 

Closes #482 

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

File: experiments/AMIP/modular/components/atmosphere/climaatmos_init.jl [hunk 1]
```diff
@@ -8,8 +8,8 @@ import ClimaCoupler.FluxCalculator:
     extrapolate_ρ_to_sfc,
     get_surface_params
 using ClimaCore: Fields.level, Geometry
-import ClimaCoupler.FieldExchanger: get_thermo_params
-import ClimaCoupler.Interfacer: get_field, update_field!, name, get_model_state_vector
+import ClimaCoupler.Interfacer: get_field, update_field!, name
+import ClimaCoupler.Checkpointer: get_model_state_vector
```
[the remark this answers is not in the document]
Comment (LenkaNovak) [reply experiments/AMIP/modular/components/atmosphere/climaatmos_init.jl:12]: This is just cleaning up wrong imports from past merges.

File: src/TimeManager.jl [hunk 2]
```diff
@@ -63,4 +72,69 @@ Returns `true` if the current date is equal to or exceeds the saved first of the
...
+"""
+do_nothing(::Interfacer.CoupledSimulation, _) = nothing
+
+"""
+    HourlyCallback{FT}
+
+This is a callback type that triggers at intervals of 1h or multiple hours.
+"""
+@kwdef struct HourlyCallback{FT} <: CouplerCallback
+    dt::FT = FT(1) # hours
```
Comment (akshaysridhar) [inline src/TimeManager.jl:91]: We should use `DocStringExtensions` and add descriptions so the API docs automatically pick up property information.

File: src/TimeManager.jl [hunk 3]
```diff
@@ -63,4 +72,69 @@ Returns `true` if the current date is equal to or exceeds the saved first of the
...
+    data::Array = []
+end
+
+"""
+    MonthlyCallback{FT}
+
+This is a callback type that triggers at intervals of 1 month or multiple months.
+"""
+@kwdef struct MonthlyCallback{FT} <: CouplerCallback
+    dt::FT = FT(1) # months
```
Comment (akshaysridhar) [inline src/TimeManager.jl:104]: Same here w.r.t `DocStringExtensions`

File: test/time_manager_tests.jl [hunk 4]
```diff
@@ -65,6 +67,115 @@ end
...
+    @test cs.callbacks.monthly_counter.data[1] == 1
+
+    cs.dates.date .+= Dates.Hour(2)
+    TimeManager.trigger_callback!(cs, cs.callbacks.twhohourly_inactive)
+    TimeManager.trigger_callback!(cs, cs.callbacks.twhohourly_nothing)
+    TimeManager.trigger_callback!(cs, cs.callbacks.twhohourly_counter)
+    TimeManager.trigger_callback!(cs, cs.callbacks.monthly_counter)
+    @test cs.callbacks.twhohourly_inactive.ref_date[1] == date0
+    @test cs.callbacks.twhohourly_nothing.ref_date[1] == cs.dates.date0[1] + Dates.Hour(4)
+    @test cs.callbacks.twhohourly_counter.ref_date[1] == cs.dates.date0[1] + Dates.Hour(4)
```
Comment (akshaysridhar) [inline test/time_manager_tests.jl:133]: Duplicate code ?

**Review by akshaysridhar — APPROVED** [review]
LGTM - Suggest adding the DocStringExtensions Fields information to the new structs so that they'll be captured in the API docbuild. I'm not sure what the solution to this windows-ci issue is (please also squash commits) .

**Pull request merged**